app/helpers/i18n_helper.rb

```diff
@@ -1,17 +1,17 @@
 module I18nHelper
   def set_locale
-    # Save a given locale
+    # Save a given locale from params
     if params[:locale] && available_locale?(params[:locale])
       spree_current_user&.update!(locale: params[:locale])
       cookies[:locale] = params[:locale]
     end
 
     # After logging in, check if the user chose a locale before
-    if spree_current_user && spree_current_user.locale.nil? && cookies[:locale]
-      spree_current_user.update!(locale: params[:locale])
+    if current_user_locale.nil? && cookies[:locale] && available_locale?(params[:locale])
```

Comment:
I dont understand why are you mixing cookies with params in this if clause?

Reply:
did you mean available_locale?(cookies[:locale]) ?

Reply:
and the line below would be
      spree_current_user&.update!(locale: cookies[:locale])
?

Reply:
I think this needs to be extracted to a method with a good name, if my corrections are correct this would be called saveLocaleFromCookies?

Reply:
Yes, this looked really strange to me as well, but I didn't want to alter the existing logic too much. The original lines are:
```ruby
# After logging in, check if the user chose a locale before
if spree_current_user && spree_current_user.locale.nil? && cookies[:locale]
  spree_current_user.update!(locale: params[:locale])
end
```

I've essentially refactored a tiny bit, and ensured we don't persist `params[:locale]` if it's not available, but the original logic is not changed. You're right though, it looks like it's either not doing anything useful, or not doing what it's intended to do...?

Reply:
I think we need a bigger hammer on this PR. I think we must go for a bigger clean up here. This code is terrible and it's not so difficult to make it simple. Let me know, I can give it a try on Monday if you prefer.

Reply:
> I think we need a bigger hammer

:smile: 

Yeah, I was going for the soft approach here and trying to leave things intact, but lets smash things instead :+1:

I'll take another look at this `cookies[:locale]` mess.

Reply:
Yeah okay I think this cookies bit is a mess and I'm not sure the related specs are even doing anything useful either. :see_no_evil: 

In the existing code we start with this conditional: 
```ruby
spree_current_user && spree_current_user.locale.nil? && cookies[:locale]
```
This is checking a case where the user **is** logged in, there is **not** a locale set on the user record, but there **is** a locale set in cookies. I guess this scenario is: a user that as never explicitly selected a locale before has selected a locale (as a guest), then logged in. Or maybe a guest user selects a locale and then registers?

Then we do this if it's true:
```ruby
spree_current_user.update!(locale: params[:locale])
```
...persisting a locale to the user record, but from the params instead of the cookies (the value in the cookies is ignored). I'm not sure there's a case where the user is both in the process of logging in **and** the locale is being passed in the params at the same time? Anyway, we should probably be checking if `cookies[:locale]` is valid in this case, and persisting that value (not the params value) if the user doesn't have a locale already...

Reply:
> we should probably be checking if cookies[:locale] is valid in this case, and persisting that value (not the params value) if the user doesn't have a locale already...

agree. I assume the cookie is needed for guest users and we cannot get rid of it, right?

Reply:
yeah, that code looked like non-sense... but I am not sure I understand everything that is going on here.
Anyway, I think this PR is a great step forward and we can iterate from here if we find any issues.